### ISSUE TITLE
add meta information to rules

### ISSUE THREAD

**Issue by MelSumner** [ember-template-lint/ember-template-lint#943]
For each rule, some meta information is needed: 

- category (1)
- tags (0-many)

This will help with automating presets and some other future features.

**Comment by rwjblue** [comment]
Also need to add:

* short description
* link to docs page
* suggested config (e.g. should you use `true`, `'whatever'`, etc)

**Comment by rwjblue** [comment]
Suggested structure:

```
class FooRule extends Rule {
  static get meta() {
   return {
     description: 'some short sentence describing the rule',
     category: "Stylistic Issues", // "Deprecated Rules", "Possible Error", "Best Practices", 
     recommended: false,
     url: "https://github.com/ember-template-lint/ember-template-lint/blob/master/docs/rules/foo.md",
   };
  }
}
```

**Comment by rwjblue** [comment]
@bmish / @Turbo87 / @dcyriller - Thoughts?

**Comment by bmish** [comment]
Looks pretty good. Can we aim to match the eslint rule properties/structure as much as possible? That would make it easier for eslint rule authors to author these template lint rules. And perhaps even allow us to take advantage of eslint tooling like eslint-plugin-eslint-plugin which lints eslint rules.

**Comment by rwjblue** [comment]
Ya, I copied the above snippet (with small modifications) from eslint's guides.

**Comment by MelSumner** [comment]
@bmish do you think that the suggested structure from @rwjblue reflects what you had in mind by matching the eslint rule structure? I'd like to start working on the issue. Thank you!

**Comment by Turbo87** [comment]
maybe we should replace `recommended` with a more detailed data structure that also supports other presets? 🤔

**Comment by bmish** [comment]
eslint's `type` meta [property](https://eslint.org/docs/developer-guide/working-with-rules) is a useful one to include too.

Would it be worth including a `docs` object to contain some of the properties like eslint has? Wondering if we should just exactly match that or if @rwjblue prefers to flatten it?

**Comment by rwjblue** [comment]
Yeah, I vaguely preferred to flatten it, but I don't really mind if we just use their exact schema either.

**Comment by mansona** [comment]
Just to catch up with this issue, does your proposal @rwjblue mean that ember-template-lint currently doesn't have a way to provide a link back to docs? 🤔 

I have been looking at the docs for `log()` and some of the source code and I don't see anything that seems to be providing a url that can be consumed by editors.

**Comment by bmish** [comment]
Currently there's no way for rules to provide a link to their doc, although I'd love it if we followed ESLint's lead and implemented this.

**Comment by mansona** [comment]
So I managed to revive @MelSumner 's work on this from before in https://github.com/ember-template-lint/ember-template-lint/pull/2070 but I'm not 100 sure that the code now matches the way the discussion here went 🤔 

The work in the PR has a meta with the structure: 

```
{
  description: 'disallows `{{debugger}}` in templates',
  category: 'Best Practices', // 'Stylistic Issues', 'Deprecated Rules', 'Possible Error', 'Best Practices',
  presets: { recommended: true },
  url: 'https://github.com/ember-template-lint/ember-template-lint/blog/master/docs/rules/no-debugger.md',
  fixable: false,
}
```

but eslint's meta looks more like this: 

```js
{
        type: "suggestion",

        docs: {
            description: "disallow unnecessary semicolons",
            category: "Possible Errors",
            recommended: true,
            url: "https://eslint.org/docs/rules/no-extra-semi"
        },
        fixable: "code",
        schema: [] // no options
    }
```

Also I'm not entirely sure why the code is using a static getter for the meta on the rule 🤔  I'm sure there was a reason but I've never used them before so I don't know the benefits 😂 

any thoughts?

**Comment by scalvert** [comment]
I think this is an opportune time to explore the potential convergence of eslint and this repo. The latest versions of eslint added the ability to [add custom parsers](https://eslint.org/docs/developer-guide/working-with-custom-parsers), which would allow us to leverage a lot of the goodness of eslint to lint handlebars templates. This is useful because it allows us to

- leverage existing infrastructure
- build rules using a well established structure

@rwjblue and I have discussed this a number of times lately, and I'd be in favor of that exploration happening in lieu of this work.

**Comment by MelSumner** [comment]
Adding meta info to this repo is a very different issue than merging eslint and ember-template-lint, IMO. Adding the meta info was, at least when I worked on it originally, intended to provide a straight-forward way to build a website for ember-template-lint.

**Comment by Turbo87** [comment]
I tend to agree that we should probably not conflate the two things, otherwise adding metadata would be blocked on this significantly bigger task. It does make sense though to keep our data structured in a way that is compatible, or at least close, to the format that ESLint uses, in case we ever want to perform this migration.

regarding the "using ESLint for template linting": for the linting itself this will probably work okay, but I'm slightly worried about supporting `--fix` with the ESLint approach. I don't know the internals that well, but last I looked it was using a very basic string-based approach, while template-lint is using template-recast.

**Comment by bmish** [comment]
I'm also generally supportive of the long-term direction of rebuilding this plugin as an ESLint plugin with a custom parser (or at least investigating the feasibility of this, @Turbo87 makes a good point about the differing approaches to autofixing). But given how much work that could be and our lack of resources, I agree that we shouldn't block incremental improvements made to the existing plugin on it, although we should try to conform to ESLint's API as much as possible.

**Comment by scalvert** [comment]
I think we're aligned on the fact that we should ideally maintain the same or a similar structure to eslint, which would make future migrations less painful. I'd be supportive of moving forward with that in mind.

And yes, @Turbo87, I believe you're right about eslint's fixer capabilities. Using recast is imo a far superior mechanism, otherwise it's just a glorified search & replace.

**Comment by bmish** [comment]
I found some old discussions about improving autofixing in ESLint (likely to more of an AST-based approach):

* https://github.com/eslint/eslint/issues/11494
* https://github.com/eslint/eslint/issues/5329

But it doesn't seem likely for improvements to be made unless someone brings them up again and champions them through the RFC process.
